### ISSUE TITLE
Cannot execute cells from keyboard shortcuts

### ISSUE THREAD

**Issue by martinRenou** [jupyterlab/jupyterlab#7412]
## Description

I cannot execute Notebook cells with `shift + enter` of `ctrl + enter`. And I get this warning in the console:

```
Cannot execute key binding 'Ctrl Enter': command 'notebook:run-cell' is not enabled
```

## Reproduce

I don't really know how to reproduce it, it seems like at some point keyboard shortcuts stop working.

## Context

- Operating System and version: Ubuntu 18.04.3 LTS
- Browser and version: Chromium Version 77.0.3865.90 (Official Build) Built on Ubuntu
- JupyterLab version: 1.1.4

**Comment by jasongrout** [comment]
I've seen this sometimes when running on master, but it was very sporadic and restarting fixed it. I didn't trace down why it was happening.

I presume it has to do with the keybinding selector somehow being satisfied: 

https://github.com/jupyterlab/jupyterlab/blob/c9d9a3978371687aecd691e4d26464560270868f/packages/notebook-extension/schema/tracker.json#L155-L164

but the command itself not being enabled, which is defined here: https://github.com/jupyterlab/jupyterlab/blob/c9d9a3978371687aecd691e4d26464560270868f/packages/notebook-extension/src/index.ts#L935-L947 which calls the isEnabled function here: https://github.com/jupyterlab/jupyterlab/blob/c9d9a3978371687aecd691e4d26464560270868f/packages/notebook-extension/src/index.ts#L893-L901

**Comment by martinRenou** [comment]
Maybe there is a keyboard shortcut that I unintentionally press that disables those commands?

**Comment by jasongrout** [comment]
I think the key will be seeing what that tracker.currentWidget is, and why the shell current widget isn't it, when clearly the focus is on something inside the notebook because the keystrokes got trapped there.

**Comment by jasongrout** [comment]
My guess is that there is some way to focus something inside the notebook without the focus bubbling up for the tracker to make the notebook the current widget.

**Comment by shimminj** [comment]
I'm seeing this sporadically on JupyterLab 1.1.4, Chrome, on OSX.

Restarting the kernel does not clear the problem..  Reconnecting in a new browser window does.

Seeing errors like:
`Cannot execute key binding 'Enter': command 'notebook:enter-edit-mode' is not enabled. 3 index.js:504`
`Cannot execute key binding 'Ctrl Enter': command 'notebook:run-cell' is not enabled. index.js:504`
`Cannot execute key binding 'Shift Enter': command 'notebook:run-cell-and-select-next' is not enabled. 7 index.js:504`

Note that 'Shift Enter' does work when the focus is outside any cell (e.g. if we click outside the cell window and there is no active cursor), but the other commands do not.

When this happens, is there any debugging we can do to get more data to help track this problem down?

**Comment by mhenault1** [comment]
I'm experiencing the same issue on jupyterlab v1.2.6, on both OSX/Ubuntu 16.04 and Chrome/Firefox. On the Chrome console, I am getting the same warning as @martinRenou:
```
Cannot execute key binding 'Ctrl Shift -': command 'notebook:split-cell-at-cursor' is not enabled.
```
This happens whatever command I try to use (run-cell, merge-cells, split-cell-at-cursor, etc). Restarting notebook, refreshing or restarting browser won't solve the problem.

It appears to be browser-specific though, as shortcuts do work on Ubuntu's default browser.

**Comment by pulkin** [comment]
I have the same issue:
```
Cannot execute key binding 'Shift Enter': command 'notebook:run-cell-and-select-next' is not enabled. 3 index.js:504
Cannot execute key binding 'Ctrl Enter': command 'notebook:run-cell' is not enabled. index.js:504
Cannot execute key binding 'Shift Enter': command 'notebook:run-cell-and-select-next' is not enabled. index.js:504
Cannot execute key binding 'Alt Enter': command 'notebook:run-cell-and-insert-below' is not enabled. index.js:504
```
I was editing a relatively big cell where the problem appeared first. Restart helped the first time but right now restart has no effect and the problem persists. Swapping the browser (chromium into Firefox) does not help either.

**Comment by pulkin** [comment]
Removing everything (except `jupyter_notebook_config.py`) from `.jupyter` helped for the moment. This issue gets 9/10 frustration score.

**Comment by jasongrout** [comment]
In these cases where people are seeing this issue, does the notebook tab have the little overline indicating that it is the current document? Also, please always tell us what jlab version and what browser version you have. And please try it in incognito/private mode to see if there is a browser plugin that might be messing things up.

**Comment by pulkin** [comment]
> In these cases where people are seeing this issue, does the notebook tab have the little overline indicating that it is the current document?

Yes. -1 frustration score because I can actually select and execute the specific cell from the drop-down main menu.

```bash
jupyter-lab --version
1.1.4
chromium-browser --version
Chromium 79.0.3945.130 Fedora Project
firefox --version
Mozilla Firefox 74.0
```

I use jupyter-lab with app mode in chromium.

**Comment by jasongrout** [comment]
If I recall correctly, some focus issues have been fixed in later versions of JupyterLab. If you can still reproduce the issue, does upgrading to 2.0 or at least the latest 1.2 version fix the issue?

**Comment by pulkin** [comment]
Unfortunately, I cannot provide more information that that. My general sense tells me that given this message in the console
```
Cannot execute key binding 'Shift Enter': command 'notebook:run-cell-and-select-next' is not enabled. 3 index.js:504
```
it is likely **not** a focus issue.

**Comment by jasongrout** [comment]
You can see in my comments above in this thread looking at the code why I think it might be a problem with tracking focus

**Comment by jasongrout** [comment]
But if there is something else you are seeing that makes it unlikely to be a focus tracking issue, definitely let us know.

**Comment by pulkin** [comment]
I think I can confirm that **it is a focus issue**. It happened once more and I managed to resolve it by opening a new tab, switching to it and back.

**Comment by jasongrout** [comment]
Can you tell us exactly what you were doing immediately before the problem? We're at the stage where I think it would be easy to track down if we could reproduce it.

Did you upgrade your JupyterLab?

**Comment by pulkin** [comment]
Yes, I upgraded `jupyterlab` after the issue occurred. I was scrolling, editing. Probably, the last thing I did was scrolling to the top of the notebook and making some edits in one of the first cells.

The best guess I can give you is that I have a particularly large cell and it first happened when I was adding even more lines to it.

**Comment by jasongrout** [comment]
What version do you have now?

**Comment by pulkin** [comment]
Right now it is 2.1.0 and I did not have any issues. If I observe the same issue I will try to save it for you to debug.

**Comment by jasongrout** [comment]
Thanks!

**Comment by Canadauni** [comment]
I've had this happen intermittently in version 1.2.6 in whatever the latest version of Firefox is. Generally it happens on the first load of my environment. I found that the tab has the active line on it however I have to actively still click the tab of the notebook for keyboard shortcuts to be active.

**Comment by nmaruthi** [comment]
I am facing a similar issue as well:

JupyterLab 2.0.0 on Windows 10, Conda 4.8.3, Chrome Version 84.0.4147.105 (Official Build) (64-bit). I am running JupyterLab on http://localhost:8888/lab and NB on http://localhost:8888/notebooks

When I open the browser console, I see this message:
![image](https://user-images.githubusercontent.com/22496243/90475402-0260f100-e0ed-11ea-8fae-44cb589a53ee.png)

I have not made any changes to the settings - all shortcuts are working on Jupyter notebook, but not on Jupyterlab

**Comment by DanielMarchand** [comment]
I also ran into this problem on Jupyter-lab 1.0.4, Chrome: Version 81.0.4044.138 (Official Build) (64-bit), Ubuntu 18. 

For some reason switching the browser to firefox solves it?

**Comment by jackmuskopf** [comment]
Also having this issue with same console log

**Comment by pierreolivierbonin** [comment]
I think it's worth sharing that there is a correct (and incorrect) and specific way to write custom keyboard shortcuts. In my case, the issue I had was that the custom keyboard shortcut I had created was appearing in the dropdown "Run" menu (and functioned properly via point and click) but was not functioning when I used the keyboard to execute it. I have found that I had incorrectly written the shortcut as follows:

...
"keys": [
                "Ctrl+Shift+Enter"
            ],
...

The reason why I had written the code snippet above was that I had simply followed the logic of what is displayed when we click on the dropdown "Run" menu in Jupyterlab. However, this was incorrect. The correct way to write a shortcut is without operators (e.g. '+'), as follows:

...
"keys": [
                "Ctrl Shift Enter"
            ],
...

which resolved the issue for me. Using Jupyterlab 2.2.6 on a Google Chrome browser.
